File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.cpp
#include <bits/stdc++.h>

using namespace std;

bool solution(string s)
{
    bool answer = true;

    stack <char> st;
    for(auto it : s) {
        if(it == '(') {
            st.push(it);
        } else {
            if(st.empty()) {
                answer = false;
                break;
            } else {
                st.pop();
            }
        }
    }
    
    if(!st.empty())
        answer = false;
    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s) {
    let answer = true;

    const stack = [];
    for (const it of s) {
        if (it === '(') {
            stack.push(it);
        } else {
            if (stack.length === 0) {
                answer = false;
                break;
            } else {
                stack.pop();
            }
        }
    }

    if (stack.length !== 0) {
        answer = false;
    }
    return answer;
}

// 테스트
const result = solution("((())())");
console.log(result); // true
